### PR TITLE
layers: Add VK_KHR_shader_float_controls2 support

### PR DESCRIPTION
**Pull request by spencer-lunarg**
Add support for `VK_KHR_shader_float_controls2`

- VUID-RuntimeSpirv-shaderSignedZeroInfNanPreserveFloat16-09559
- VUID-RuntimeSpirv-shaderSignedZeroInfNanPreserveFloat16-09560
-  VUID-RuntimeSpirv-shaderSignedZeroInfNanPreserveFloat32-09561
-  VUID-RuntimeSpirv-shaderSignedZeroInfNanPreserveFloat32-09562
-  VUID-RuntimeSpirv-shaderSignedZeroInfNanPreserveFloat64-09563
-  VUID-RuntimeSpirv-shaderSignedZeroInfNanPreserveFloat64-09564

### CODE REVIEW

File: layers/core_checks/cc_spirv.cpp [hunk 1]
```diff
@@ -1070,93 +1070,192 @@ bool CoreChecks::ValidateAtomicsTypes(const spirv::Module &module_state, const s
...
+            const bool has_mask = (fast_math_mode->GetConstantValue() & mask) == mask;
+            if (has_mask) {
+                continue;  // nothing to validate
+            }
+
+            const uint32_t bit_width = get_float_width(insn->Word(3));
+
+            if (bit_width == 16 && !phys_dev_props_core12.shaderSignedZeroInfNanPreserveFloat16) {
+                skip |= LogError("VUID-RuntimeSpirv-shaderSignedZeroInfNanPreserveFloat16-09559", module_state.handle(), loc,
+                                 "shaderSignedZeroInfNanPreserveFloat16 is false, but set 16-bit floats with FPFastMathDefault "
```
Comment (arno-lunarg) [inline layers/core_checks/cc_spirv.cpp:1183]: I think that the subject of `set` is not clear. Could you rephrase that please?

File: layers/core_checks/cc_spirv.cpp [hunk 2]
```diff
@@ -1070,93 +1070,192 @@ bool CoreChecks::ValidateAtomicsTypes(const spirv::Module &module_state, const s
...
+        for (; operand_index < target_insn->Length(); operand_index++) {
+            const uint32_t bit_width = get_float_width(target_insn->Word(operand_index));
+            float_16 |= bit_width == 16;
+            float_32 |= bit_width == 32;
+            float_64 |= bit_width == 64;
+        }
+
+        if (float_16 && !phys_dev_props_core12.shaderSignedZeroInfNanPreserveFloat16) {
+            skip |= LogError("VUID-RuntimeSpirv-shaderSignedZeroInfNanPreserveFloat16-09560", module_state.handle(), loc,
+                             "shaderSignedZeroInfNanPreserveFloat16 is false, but FPFastMathMode with mode 0x%" PRIx32
```
Comment (arno-lunarg) [inline layers/core_checks/cc_spirv.cpp:1235]: Sentence is weird here too I think (usage of `but`)

File: tests/unit/shader_spirv.cpp [hunk 3]
```diff
@@ -2148,3 +2148,95 @@ TEST_F(NegativeShaderSpirv, InvalidExtension) {
...
+    // Missing NotNaN
+    const char *spv_source = R"(
+        OpCapability Shader
+        OpCapability FloatControls2
+        OpExtension "SPV_KHR_float_controls2"
+        OpMemoryModel Logical GLSL450
+        OpEntryPoint GLCompute %main "main"
+        OpExecutionModeId %main FPFastMathDefault %float %constant
+        OpExecutionMode %main LocalSize 1 1 1
+        OpDecorate %add FPFastMathMode NSZ|NotInf|NotNaN
```
Comment (arno-lunarg) [inline tests/unit/shader_spirv.cpp:2175]: Maybe I just do not read this correctly, but `NotNaN` is present in this mask?

**Review by arno-lunarg — COMMENTED** [review]
Error messages could maybe be reworked, and I think there is an issue in a test

File: tests/unit/shader_spirv.cpp (hunk 3 above)
Comment (spencer-lunarg) [reply tests/unit/shader_spirv.cpp:2175]: yes, but this test is for the `FPFastMathDefault`

the below test is for `FPFastMathMode`

File: layers/core_checks/cc_spirv.cpp (hunk 1 above)
Comment (spencer-lunarg) [reply layers/core_checks/cc_spirv.cpp]: fixed

File: tests/unit/shader_spirv.cpp [hunk 4]
```diff
@@ -2148,3 +2148,95 @@ TEST_F(NegativeShaderSpirv, InvalidExtension) {
...
+    AddRequiredFeature(vkt::Feature::shaderFloatControls2);
+    RETURN_IF_SKIP(Init());
+
+    VkPhysicalDeviceFloatControlsProperties shader_float_control = vku::InitStructHelper();
+    GetPhysicalDeviceProperties2(shader_float_control);
+    if (shader_float_control.shaderSignedZeroInfNanPreserveFloat32) {
+        GTEST_SKIP() << "shaderSignedZeroInfNanPreserveFloat32 is supported";
+    }
+
+    // Missing NotNaN
```
Comment (arno-lunarg) [inline tests/unit/shader_spirv.cpp:2166]:
```suggestion
```

**Review by arno-lunarg — APPROVED** [review]
